balloc/balloc.c

```diff
@@ -2251,6 +2252,29 @@ static int balloc_wild_scan_group(struct balloc_allocation_context *bac,
 	list = &grp->bgi_normal.bzp_extents;
 #endif
 
+	/**
+	 * Check to detect the block allocation request which came earlier
+	 * for criteria 1 on same group and come again for criteria 2 with
+	 * already set best extent.
+	 * In criteria 1 bac_status not marked to M0_BALLOC_AC_FOUND
+	 * because bac_found < M0_BALLOC_DEFAULT_MIN_TO_SCAN
+	 * Now with criteria 2 on same group with already set best extent may
+	 * not part of extent in free extent list because another requests
+	 * may have updated extents in free list.
+	 * Reset best extent by detecting this case so that it
+	 * will find correct best extent, also set end_of_group flag so
+	 * balloc_check_limits() could call balloc_use_best_found() to set
+	 * final extent from best extent.
+	 */
+	if (bac->bac_found != 0) {
+		m0_bindex_t group = balloc_bn2gn(bac->bac_best.e_start,
+						 bac->bac_ctxt);
+		if (group == grp->bgi_groupno) {
+			end_of_group = 1;
+			M0_SET0(&bac->bac_best);
+			m0_ext_init(&bac->bac_best);
```

Comment:
As I said in another comment, I think we should reset this in balloc_regular_allocator()
when lock is released and is going to try another criteria.

When lock is released, other threads might get chance to allocate block from the group where the "bac_best" belongs to.

Reply:
>>When lock is released, other threads might get chance to allocate block from the group where the "bac_best" belongs to.

1.  other thread allocating blocks belongs to bac_best which is ok
2.  thread goes in criteria2 with older bac_best (which is updated in step1) but for different group will not go for DB update because of current implementation of wild scan, we set final extent when bac_best belongs to same group only.
request continue in loop till it get same group and there we clear the bac_best.

so this case handled here.
Purpose to move fix from balloc_regular_allocator() to wild scan is due to keeping readability simpler.